DLPX-83442 Disable various kernel modules which we don't use

prakashsurya opened the pull request.

sebroy commented on debian.aws/config/OVERRIDES:
```diff
@@ -0,0 +1,57 @@
+CONFIG_DRM=n
+CONFIG_INFINIBAND=n
+CONFIG_NFC=n
+CONFIG_USB=n
+CONFIG_USB=n
```
Looks like a dup (we _really_ don't want USB!) ;-)

sebroy reviewed (commented):
Is there a way to pass in these overrides as parameters to the build job? I'm just exploring whether there's a possibility of maintaining this list in a single place for all kernels as opposed to duplicating it in all of the kernel repos... (?)

prakashsurya commented:
> Is there a way to pass in these overrides as parameters to the build job?

Hm.. I'm sure we could dynamically copy it into place as part of the linux-pkg build.. I do think it's a bit awkward to duplicate it in all our platform specific repositories, but at the same time, we do that already for all other kernel changes we make.. so I'm not sure if this specific change should be treated any differently.. but, perhaps we should.. I'm not sure I have an opinion yet, I see both sides having merit (duplicate in all kernel repos, or not)..

I'll think about this some more...

prakashsurya commented:
Finally got a build to work...

Before:
```
delphix@ip-10-110-235-80:~$ dpkg-query -L linux-modules-5.4.0-1085-dx2022092505-805ba8691-aws | grep \.ko$ | wc -l
943
delphix@ip-10-110-235-80:~$ dpkg-query -L linux-modules-extra-5.4.0-1085-dx2022092505-805ba8691-aws | grep \.ko$ | wc -l
2704
```

After:
```
delphix@ip-10-110-230-45:~$ dpkg-query -L linux-modules-5.4.0-1085-dx2022101118-e92c3e5a4-aws | grep \.ko$ | wc -l
926
delphix@ip-10-110-230-45:~$ dpkg-query -L linux-modules-extra-5.4.0-1085-dx2022101118-e92c3e5a4-aws | grep \.ko$ | wc -l
2568
```

Not a huge difference, but still something... probably more of a difference in the size, as it looks like we've removed some of the biggest modules..

Before:
```
$ find /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws -type f -name '*.ko' | xargs du -b | sort -n | tail
1517305 /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko
1651089 /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/fs/ocfs2/ocfs2.ko
1856433 /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/fs/cifs/cifs.ko
2046881 /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/fs/xfs/xfs.ko
2126433 /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/fs/btrfs/btrfs.ko
2315177 /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/drivers/gpu/drm/radeon/radeon.ko
3340993 /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/drivers/gpu/drm/i915/i915.ko
3543641 /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/kernel/kheaders.ko
6170352 /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/extra/zfs.ko
7149665 /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/drivers/gpu/drm/amd/amdgpu/amdgpu.ko
```

After:
```
$ find /usr/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel -type f -name '*.ko' | xargs du -b | sort -n | tail
830921  /usr/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/drivers/net/ethernet/qlogic/qed/qed.ko
1083481 /usr/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/arch/x86/kvm/kvm.ko
1100985 /usr/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/fs/nfs/nfsv4.ko
1141993 /usr/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/drivers/net/ethernet/mellanox/mlxsw/mlxsw_spectrum.ko
1150657 /usr/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/drivers/net/ethernet/broadcom/bnx2x/bnx2x.ko
1183185 /usr/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/drivers/scsi/qla2xxx/qla2xxx.ko
1456257 /usr/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/drivers/scsi/lpfc/lpfc.ko
1516457 /usr/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko
1856433 /usr/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/fs/cifs/cifs.ko
3536985 /usr/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/kernel/kheaders.ko
```

sdimitro commented:
@prakashsurya let's make the arguments for this PR more compelling to others :) - above you say that we removed some of the biggest modules but you showcase the directory that has the stripped binaries like this:
```
delphix@sd-DLPX-60981:~$ find /lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/ -type f -name '*.ko' | xargs du -b | sort -n | tail
1707441	/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/lib/test_bpf.ko
1856497	/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/fs/cifs/cifs.ko
2046881	/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/fs/xfs/xfs.ko
2126433	/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/fs/btrfs/btrfs.ko
2315177	/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/drivers/gpu/drm/radeon/radeon.ko
3204409	/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/drivers/gpu/drm/nouveau/nouveau.ko
3340929	/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/drivers/gpu/drm/i915/i915.ko
3583897	/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/kernel/kheaders.ko
6170352	/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/extra/zfs.ko
7149681	/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/drivers/gpu/drm/amd/amdgpu/amdgpu.ko
```
At that point we see that the amdgpu.ko module takes around 7MB, but that's not the whole story. If you look at the directory with the debug info like this:
```
delphix@sd-DLPX-60981:~$ find /usr/lib/debug/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic -type f -name '*.ko' | xargs du -b | sort -n | tail
34235409	/usr/lib/debug/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/drivers/infiniband/hw/hfi1/hfi1.ko
36567320	/usr/lib/debug/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/drivers/staging/rtl8723bs/r8723bs.ko
41266553	/usr/lib/debug/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/net/mac80211/mac80211.ko
47682657	/usr/lib/debug/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/fs/xfs/xfs.ko
53728001	/usr/lib/debug/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/drivers/gpu/drm/radeon/radeon.ko
78037769	/usr/lib/debug/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko
116411752	/usr/lib/debug/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/extra/zfs.ko
128963817	/usr/lib/debug/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/drivers/gpu/drm/i915/i915.ko
222555153	/usr/lib/debug/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/drivers/gpu/drm/nouveau/nouveau.ko
226726577	/usr/lib/debug/lib/modules/5.4.0-126-dx2022092505-cd7eac0b7-generic/kernel/drivers/gpu/drm/amd/amdgpu/amdgpu.ko
```
You can see that the same module with its debug info that's part of our images are 226MB, and the second one is 222MB (which we also strip). I'd guess that with your changes we save more than 1 GB in debug info and probably close to 50~100MB of unstripped data.

Other points include:
* Saving memory in our crash kernel - the crash kernel doesn't have to load all these modules, and its always loaded up in memory to get crash dumps. With this change you help us lower its memory consumptions and stirs us further away from OOM bugs when capturing crash dumps
* The storage saved above is not just from the rpool, but also from the images that we sent to our customers, including upgrade images.
* Appliance build time - we don't spend time building all these big graphics cards and Ubuntu's ZFS module. (Do you have any idea how much time we save there?)

prakashsurya commented:
@sdimitro yea, for sure.. you're definitely right..

Before:
```
$ find /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws -type f -name '*.ko' | xargs du -b | sort -n | tail
27737817        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/fs/cifs/cifs.ko
34217553        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/fs/btrfs/btrfs.ko
34224025        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/drivers/infiniband/hw/hfi1/hfi1.ko
41256625        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/net/mac80211/mac80211.ko
47670969        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/fs/xfs/xfs.ko
53705529        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/drivers/gpu/drm/radeon/radeon.ko
78012961        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko
116410920       /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/extra/zfs.ko
128893905       /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/drivers/gpu/drm/i915/i915.ko
226634721       /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws/kernel/drivers/gpu/drm/amd/amdgpu/amdgpu.ko
```

After:
```
$ find /usr/lib/debug/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws -type f -name '*.ko' | xargs du -b | sort -n | tail
15955793        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/fs/nfsd/nfsd.ko
16613697        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/net/sunrpc/sunrpc.ko
17090129        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/net/rxrpc/rxrpc.ko
18339633        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/fs/nfs/nfsv4.ko
20836289        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/net/sctp/sctp.ko
26590161        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/drivers/net/ethernet/netronome/nfp/nfp.ko
26922337        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/drivers/net/ethernet/mellanox/mlxsw/mlxsw_spectrum.ko
27724657        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/fs/cifs/cifs.ko
77962217        /usr/lib/debug/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko
116386800       /usr/lib/debug/lib/modules/5.4.0-1085-dx2022101118-e92c3e5a4-aws/extra/zfs.ko
```

> Saving memory in our crash kernel 

:+1: agreed.

> The storage saved above is not just from the rpool, but also from the images that we sent to our customers, including upgrade images.

yea, for sure.. and it's a 5x savings here, since we have a kernel package for each platform, all bundled into a single upgrade image.. so, eliminating a large binary in the kernel, will remove it from all 5 kernel packages for our 5 current platforms..


> Appliance build time - we don't spend time building all these big graphics cards and Ubuntu's ZFS module. (Do you have any idea how much time we save there?)

I think you mean, kernel build time, right? i.e. the time it takes to build the kernel package.. since the appliance build, just consumes the pre-built kernel package.. we still may save time in the appliance build, though, but that'll be due to a smaller package size, not compile time savings..

prakashsurya commented:
@sdimitro also something I was curious about.. there's a "linux-modules-extra" package, that contains lots of kernel modules.. I'm curious if we need to install that package on our appliance? do you know?

I've opened this: https://github.com/delphix/delphix-kernel/pull/20 to see if maybe we can remove that package wholesale.. but I'm not yet sure if that's safe to do..

From what I can tell, any module that's built, but not in the "inclusion list" will wind up in this "extra" package.. and from just a number of modules perspective, there's a lot more modules in the "extra" package, than the regular one.. e.g.
```
$ dpkg-query -L linux-modules-5.4.0-1085-dx2022101118-e92c3e5a4-aws | grep \.ko$ | wc -l
926

$ dpkg-query -L linux-modules-extra-5.4.0-1085-dx2022101118-e92c3e5a4-aws | grep \.ko$ | wc -l
2568
```
I'm still investigating if there's any modules we care about in the "extra" package, though...

prakashsurya commented:
Hm.. looks like even without the "extra" package, we still get all of the debug modules:
```
$ find /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws -type f -name '*.ko' | wc -l
3649
```
as there only seems to be a single package for the debug modules:
```
$ dpkg-query -S /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws
linux-image-5.4.0-1085-dx2022092505-805ba8691-aws-dbgsym, zfs-modules-5.4.0-1085-dx2022092505-805ba8691-aws-dbg: /usr/lib/debug/lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws
```

prakashsurya commented:
```
$ find /lib/modules/5.4.0-1085-dx2022092505-805ba8691-aws -type f -name '*.ko' | wc -l
946
```
it does significantly reduce the number of non-debug modules on the system, though..

sdimitro commented:
@prakashsurya 

>> Appliance build time - we don't spend time building all these big graphics cards and Ubuntu's ZFS module. (Do you have  any idea how much time we save there?)
> 
> I think you mean, kernel build time, right? i.e. the time it takes to build the kernel package.. since the appliance build, just consumes the pre-built kernel package.. we still may save time in the appliance build, though, but that'll be due to a smaller package size, not compile time savings..

Yeah, apologies. I meant the kernel build time.

> @sdimitro also something I was curious about.. there's a "linux-modules-extra" package, that contains lots of kernel modules.. I'm curious if we need to install that package on our appliance? do you know?

Unfortunately I don't. Looking at the file contents of the deb file it looks like there is a lot of stuff geared towards desktop users but I'd be hesitant to exclude it. My fear is that some driver may be part of this deb package that we may need to have on the VM even if we are not currently using. The best example that I have for that was the ENA network driver on AWS. We always shipped with that driver even though it wasn't used - then when customers wanted to enable the driver for faster networking, the just had to reboot their VMs - no hotfix, no upgrade involved.

BTW I noticed on your output above that we still compile and keep around Ubuntu's ZFS module in our build. Any way we can stop that from compiling?

prakashsurya commented:
> BTW I noticed on your output above that we still compile and keep around Ubuntu's ZFS module in our build. Any way we can stop that from compiling?

yea, I haven't gotten to that yet, but we definitely want to remove ZFS from the kernel package.. I just haven't spent the time to figure out where to do that.. but it's on my list..

> The best example that I have for that was the ENA network driver on AWS.

sure, but at the same time.. we'd want to weigh the benefit of removing it, with the _potential_ cost of it _maybe_ it being useful in the future.. especially considering, it'd only take an upgrade+reboot to re-instate any module that we removed..

but yea, I see your point.

prakashsurya commented:
As is, here's what I have...

***

Number of non-debug modules...

Before:
```
delphix@ip-10-110-230-118:~$ dpkg-query -L linux-modules-5.4.0-1086-dx2022101316-ab9e4e437-aws |  grep \.ko$ | wc -l
943
delphix@ip-10-110-230-118:~$ dpkg-query -L linux-modules-extra-5.4.0-1086-dx2022101316-ab9e4e437-aws |  grep \.ko$ | wc -l
2704
```

After:
```
delphix@ip-10-110-249-181:~$ dpkg-query -L linux-modules-5.4.0-1086-dx2022101718-03c1dc269-aws | grep \.ko$ | wc -l
777
delphix@ip-10-110-249-181:~$ dpkg-query -L linux-modules-extra-5.4.0-1086-dx2022101718-03c1dc269-aws | grep \.ko$ | wc -l
1167
```

So, that's a difference of 1703 modules that we've removed.

***

Same for the debug modules...

Before:
```
delphix@ip-10-110-230-118:~$ dpkg-query -L linux-image-5.4.0-1086-dx2022101316-ab9e4e437-aws-dbgsym | grep \.ko$ | wc -l
3647
```

After:
```
delphix@ip-10-110-249-181:~$ dpkg-query -L linux-image-5.4.0-1086-dx2022101718-03c1dc269-aws-dbgsym | grep \.ko$ | wc -l
1944
```

***

We can see we removed lots/all of the biggest (installed file size) modules too...

Before:
```
delphix@ip-10-110-230-118:~$ dpkg-query -L linux-image-5.4.0-1086-dx2022101316-ab9e4e437-aws-dbgsym | grep \.ko$ | xargs du -b | sort -n | tail
26994105        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101316-ab9e4e437-aws/kernel/drivers/net/ethernet/mellanox/mlxsw/mlxsw_spectrum.ko
27882729        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101316-ab9e4e437-aws/kernel/fs/cifs/cifs.ko
34239017        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101316-ab9e4e437-aws/kernel/drivers/infiniband/hw/hfi1/hfi1.ko
34484993        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101316-ab9e4e437-aws/kernel/fs/btrfs/btrfs.ko
41262489        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101316-ab9e4e437-aws/kernel/net/mac80211/mac80211.ko
47685969        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101316-ab9e4e437-aws/kernel/fs/xfs/xfs.ko
53735585        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101316-ab9e4e437-aws/kernel/drivers/gpu/drm/radeon/radeon.ko
78301449        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101316-ab9e4e437-aws/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko
129423617       /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101316-ab9e4e437-aws/kernel/drivers/gpu/drm/i915/i915.ko
227066993       /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101316-ab9e4e437-aws/kernel/drivers/gpu/drm/amd/amdgpu/amdgpu.ko
```

After:
```
delphix@ip-10-110-249-181:~$ dpkg-query -L linux-image-5.4.0-1086-dx2022101718-03c1dc269-aws-dbgsym | grep \.ko$ | xargs du -b | sort -n | tail
13150689        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/net/bridge/bridge.ko
13938249        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/net/tipc/tipc.ko
14085673        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/chelsio/cxgb4/cxgb4.ko
14091249        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/scsi/qla2xxx/qla2xxx.ko
15038441        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/fs/nfs/nfs.ko
16084617        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/fs/nfsd/nfsd.ko
16688233        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/net/sunrpc/sunrpc.ko
17046825        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/net/rxrpc/rxrpc.ko
18394145        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/fs/nfs/nfsv4.ko
20860401        /usr/lib/debug/lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/net/sctp/sctp.ko
```

***

As a result, we can see the package size is much smaller too...

Before:
```
delphix@ip-10-110-230-118:~$ dpkg-query -L linux-image-5.4.0-1086-dx2022101316-ab9e4e437-aws-dbgsym | grep \.ko$ | xargs du -k | awk '{sum += $1} END {print sum}'
2055155

delphix@ip-10-110-230-118:~$ dpkg-query -Wf '${Installed-Size}  ${Package}\n' linux-image-5.4.0-1086-dx2022101316-ab9e4e437-aws-dbgsym
4727419 linux-image-5.4.0-1086-dx2022101316-ab9e4e437-aws-dbgsym
```

After:
```
delphix@ip-10-110-249-181:~$ dpkg-query -L linux-image-5.4.0-1086-dx2022101718-03c1dc269-aws-dbgsym | grep \.ko$ | xargs du -k | awk '{sum += $1} END {print sum}'
895962

delphix@ip-10-110-249-181:~$ dpkg-query -Wf '${Installed-Size} ${Package}\n' linux-image-5.4.0-1086-dx2022101718-03c1dc269-aws-dbgsym
2331052 linux-image-5.4.0-1086-dx2022101718-03c1dc269-aws-dbgsym
```

***

> Is there a way to pass in these overrides as parameters to the build job? I'm just exploring whether there's a possibility of maintaining this list in a single place for all kernels as opposed to duplicating it in all of the kernel repos... (?)

After thinking about this a bit, I do think we _could_ add this OVERRIDES file to linux-pkg, and dynamically apply it to the kernel when we build.. but, since we currently duplicate kernel changes to all kernel repos today, I'm not sure this change should be treated any differently..

While I think it's awkward to duplicate things in all of our platform specific kernel repositories, I think being consistent with how we make other changes to the kernel repos makes more sense to me.. than trying to eliminate the duplication only for this change, but not for any other changes we make to the kernels.

If folks disagree, though, we can discuss this some more...

***

@sebroy @sdimitro how do you feel about moving forward with this as-is, given some of the data I posted above? I'm not sure what the target should be for this work, w.r.t. how many modules we'd want to remove before we think we're "done" and/or removed all of the ones that we don't need..

prakashsurya commented:
If we want to try and slim things some more, I could probably go through what's left in the "linux-modules-extra" package, and find the corresponding config options to remove those too.. e.g. starting with the biggest ones..
```
delphix@ip-10-110-249-181:~$ dpkg-query -L linux-modules-extra-5.4.0-1086-dx2022101718-03c1dc269-aws | grep \.ko$ | xargs du -b | sort -n | tail -n 25
220313  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/scsi/libfc/libfc.ko
232065  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/misc/mic/scif/scif.ko
233185  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/scsi/csiostor/csiostor.ko
235177  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/brocade/bna/bna.ko
236729  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/scsi/aic7xxx/aic7xxx.ko
244761  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/scsi/isci/isci.ko
248521  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/chelsio/cxgb3/cxgb3.ko
266569  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/scsi/ufs/ufshcd-core.ko
275993  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/scsi/aic7xxx/aic79xx.ko
276041  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/stmicro/stmmac/stmmac.ko
297681  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/scsi/pm8001/pm80xx.ko
304161  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/sfc/falcon/sfc-falcon.ko
306001  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/cavium/liquidio/liquidio_vf.ko
318713  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/net/batman-adv/batman-adv.ko
333361  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/amd/xgbe/amd-xgbe.ko
355409  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/broadcom/bnxt/bnxt_en.ko
373297  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/cavium/liquidio/liquidio.ko
396801  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/intel/ice/ice.ko
485793  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/scsi/qla4xxx/qla4xxx.ko
498369  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/qlogic/qlcnic/qlcnic.ko
511385  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/sfc/sfc.ko
555737  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/target/iscsi/iscsi_target_mod.ko
664361  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/net/ethernet/chelsio/cxgb4/cxgb4.ko
754489  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/scsi/bfa/bfa.ko
3499417 /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/kernel/kheaders.ko
```

prakashsurya commented:
Also.. looking at the jenkins job times, I think the time to build the kernel package has gone from about 1 hour and 15 mins.. to about 38 mins.. so, nearly halved.

sdimitro commented:
@prakashsurya that's awesome about the kernel build times!!

As for the extras package I'd be very cautious - just glancing on your output above I found the iscsi_target module:
```
555737  /lib/modules/5.4.0-1086-dx2022101718-03c1dc269-aws/kernel/drivers/target/iscsi/iscsi_target_mod.ko
```
which we definitely need.

> @sebroy @sdimitro how do you feel about moving forward with this as-is, given some of the data I posted above? I'm not sure what the target should be for this work, w.r.t. how many modules we'd want to remove before we think we're "done" and/or removed all of the ones that we don't need..

I'm all for this going as-is if tests are passing :shipit:

prakashsurya commented:
> I'm all for this going as-is if tests are passing

do you have any thoughts on what tests I should use to verify this?

sdimitro commented:
This commit is for AWS so I'd say normal ab-pre-push on AWS for starters. Then Id target any blackbox tests that can cover any potential drivers that we don't use in normal dx-test. I'm having a hard time thinking of anything specific as I'd expect nfs and iscsi are covered in the standard BB tests.

sdimitro commented on debian.aws/config/OVERRIDES:
```diff
@@ -0,0 +1,237 @@
+CONFIG_EXFAT_FS=n
+CONFIG_EXT2_FS=n
+CONFIG_EXT3_FS=n
+CONFIG_EXT4_FS=n
```
Are we sure we want to get rid of ext4? I know that we don't use it for anything but I wonder if there is ever any scenario that we need it as a backup.

prakashsurya replied:
backup for what? I mean, we can include it.. but I'm not sure how/why we'd use it..

sdimitro replied:
I was curious if GRUB uses it or something for the recovery environment but I think that's ZFS all around too so I guess we are good without ext4. Dropping this.

prakashsurya replied:
looks like docker uses an ext4 mountpoint.. so I'll add ext4 back in..
```
Oct 21 19:15:31 ip-10-110-251-69 check_docker_start_deps[15599]: mount: /var/lib/docker: unknown filesystem type 'ext4'.
Oct 21 19:15:31 ip-10-110-251-69 check_docker_start_deps[15403]: /opt/delphix/server/bin/check_docker_start_deps: failed to mount docker zvol
```

sdimitro replied:
ok glad we caught this!

prakashsurya commented:
OK, sounds good.. that's what I was thinking too.. I'm getting close to having a "final" version that I'll test out..

prakashsurya commented:
I added some more config overrides, and have gotten the extras package down even further:
```
delphix@ip-10-110-230-118:~/test$ dpkg -c linux-modules-extra-5.4.0-1086-dx2022101920-46b8ccc9d-aws_5.4.0-1086.93_amd64.deb | grep \.ko$ | wc -l
989
```

and total modules:
```
delphix@ip-10-110-230-118:~/test$ dpkg -c linux-image-5.4.0-1086-dx2022101920-46b8ccc9d-aws-dbgsym_5.4.0-1086.93_amd64.ddeb | grep \.ko$ | wc -l
1755
```

I think I'll stop here, and move on to testing this...

prakashsurya commented:
@sebroy @sdimitro I think I have something that'll pass pre-checkin tests (running now).. can I get another round of reviews? also, are there any additional tests (besides the usual pre-checkin) that I should run?

prakashsurya commented:
Here's what it looks like currently, w.r.t. the number and size of the modules:
```
$ dpkg-query -L linux-image-5.4.0-1086-dx2022110117-805b06b5d-aws-dbgsym | grep \.ko$ | wc -l
1827

$ dpkg-query -L linux-image-5.4.0-1086-dx2022110117-805b06b5d-aws-dbgsym | grep \.ko$ | xargs du -k | awk '{sum += $1} END {print sum}'
754441

$ dpkg-query -L linux-image-5.4.0-1086-dx2022110117-805b06b5d-aws-dbgsym | grep \.ko$ | xargs du -k | sort -n | tail
6359    /usr/lib/debug/lib/modules/5.4.0-1086-dx2022110117-805b06b5d-aws/kernel/drivers/net/ethernet/intel/i40e/i40e.ko
6484    /usr/lib/debug/lib/modules/5.4.0-1086-dx2022110117-805b06b5d-aws/kernel/net/tipc/tipc.ko
6693    /usr/lib/debug/lib/modules/5.4.0-1086-dx2022110117-805b06b5d-aws/kernel/net/bridge/bridge.ko
6975    /usr/lib/debug/lib/modules/5.4.0-1086-dx2022110117-805b06b5d-aws/kernel/drivers/net/ethernet/chelsio/cxgb4/cxgb4.ko
7205    /usr/lib/debug/lib/modules/5.4.0-1086-dx2022110117-805b06b5d-aws/kernel/fs/nfs/nfs.ko
7484    /usr/lib/debug/lib/modules/5.4.0-1086-dx2022110117-805b06b5d-aws/kernel/fs/nfsd/nfsd.ko
7912    /usr/lib/debug/lib/modules/5.4.0-1086-dx2022110117-805b06b5d-aws/kernel/net/sunrpc/sunrpc.ko
8201    /usr/lib/debug/lib/modules/5.4.0-1086-dx2022110117-805b06b5d-aws/kernel/net/rxrpc/rxrpc.ko
8407    /usr/lib/debug/lib/modules/5.4.0-1086-dx2022110117-805b06b5d-aws/kernel/fs/nfs/nfsv4.ko
10218   /usr/lib/debug/lib/modules/5.4.0-1086-dx2022110117-805b06b5d-aws/kernel/net/sctp/sctp.ko

$ dpkg-query -Wf '${Installed-Size} ${Package}\n' linux-image-5.4.0-1086-dx2022110117-805b06b5d-aws-dbgsym
2121061 linux-image-5.4.0-1086-dx2022110117-805b06b5d-aws-dbgsym
```

sdimitro commented on debian.aws/config/OVERRIDES:
```diff
@@ -0,0 +1,234 @@
+CONFIG_INTEL_MEI=n
+CONFIG_IONIC=n
+CONFIG_JME=n
+CONFIG_KVM=n
```
Don't we need KVM for Docker?

prakashsurya replied:
I don't think so.. but let me fine out what test suite I can run to stress the docker stuff.. that way I can verify.

I think KVM is creating virtual machines on the host.. which we don't do.. since a container is different than a virtual machine.. my intention was not to modify any of the container stuff, but turn off the hypervisor functionality.

sebroy commented on debian.aws/config/OVERRIDES:
```diff
@@ -0,0 +1,234 @@
+CONFIG_AUXDISPLAY=n
+CONFIG_BCMA=n
+CONFIG_CONNECTOR=n
+CONFIG_CRYPTO_HW=n
```
I wonder whether this could inadvertently and silently hurt us. I'm no expert here, but is it possible that the presence of crypto hardware can be used to silently boost performance of crypto functions in a virtual machine, and that removing this support could disable this kind of capability?

prakashsurya replied:
is there a way to test that? alternatively, I can just leave this out, and keep using the default.. we can always re-evaluate later.. I go ahead and do that, so we can move this forward.. are you OK with that? i.e. just remove this line..

sebroy replied:
I'm not sure, it might be safer to leave `CONFIG_CRYPTO_HW` the way it was especially since it's security-related (unlikely to raise any security red flags).

prakashsurya replied:
right, I agree.. I've removed `CONFIG_CRYPTO_HW` from the override file.

sebroy commented on debian.aws/config/OVERRIDES:
```diff
@@ -0,0 +1,234 @@
+CONFIG_HWMON=n
+CONFIG_IAVF=n
+CONFIG_ICE=n
+CONFIG_INFINIBAND=n
```
dup from line 39

sebroy commented on debian.aws/config/OVERRIDES:
```diff
@@ -0,0 +1,234 @@
+CONFIG_NET_VENDOR_MYRI=n
+CONFIG_NET_VENDOR_NETERION=n
+CONFIG_NET_VENDOR_ROCKER=n
+CONFIG_NFC=n
```
dup from line 12

sebroy commented on debian.aws/config/OVERRIDES:
```diff
@@ -0,0 +1,234 @@
+CONFIG_SSB=n
+CONFIG_STMMAC_ETH=n
+CONFIG_SXGBE_ETH=n
+CONFIG_SYNCLINK=n
+CONFIG_SYNCLINK=n
```
duplicates

sebroy reviewed (approved):
Testing-wise, it'd be good to run this through the self-service performance testing service. Also, let's check some AWS-specific things that aren't covered by automated testing that might depend on kernel functionality, like the virtual serial console support in EC2.

prakashsurya commented:
- `virtualization_sdk` test run [here](http://selfservice.jenkins.delphix.com/job/blackbox-self-service/56142/consoleFull)

prakashsurya commented:
Attempting to do a performance test run:
```
± ./QARunner -qqq appliance -i ps-kmin-perf.dlpxdc.co file remote --run-params='-v -l debug -p use_object_store=True -p object_store_key_access_type=AWS -p hardware_config=gp3_2x16K -p reporting.report-to-dashboard=true -p reporting.influx.db_name=DOSE'
Your build is at: http://selfservice.jenkins.delphix.com/job/blackbox-self-service/56148/
```

prakashsurya commented:
> like the virtual serial console support in EC2.

@sebroy i.e. check that `dc instance console` still works? seems like it does.. e.g.
```
$ dc instance console ps-kmin-perf                                                                                                                                                                                                                                                                                                                                                                                                          
Nov  7 17:15:04 cloud-init: #############################################################                                                                                                                                                                                                                                                                                                                                                  
<14>Nov  7 17:15:04 cloud-init: -----BEGIN SSH HOST KEY FINGERPRINTS-----                                                                                                                                                                                                                                                                                                                                                                  
<14>Nov  7 17:15:04 cloud-init: 1024 SHA256:js0LV9XmkouqaOwbSJHi5q/0QHH7m8NJw6Cv+JV9o98 root@ip-10-110-197-191 (DSA)                                                                                                                                                                                                                                                                                                                       
<14>Nov  7 17:15:04 cloud-init: 256 SHA256:TiXE4licwQAfktf1KWJOMc02AvRuYZ8PQwDlF2az6zU root@ip-10-110-197-191 (ECDSA)                                                                                                                                                                                                                                                                                                                      
<14>Nov  7 17:15:04 cloud-init: 256 SHA256:WPcaIs+kpdK34GjLqfTKX+FobRr4+La7jRKiRwMvzkc root@ip-10-110-197-191 (ED25519)                                                                                                                                                                                                                                                                                                                    
<14>Nov  7 17:15:04 cloud-init: 3072 SHA256:dwGL0jAY/o18RNZrCvzLJnOmx1T0Y1x0Z+3bEdD92Og root@ip-10-110-197-191 (RSA)                                                                                                                                                                                                                                                                                                                       
<14>Nov  7 17:15:04 cloud-init: -----END SSH HOST KEY FINGERPRINTS-----                                                                                                                                                                                                                                                                                                                                                                    
<14>Nov  7 17:15:04 cloud-init: #############################################################                                                                                                                                                                                                                                                                                                                                              
-----BEGIN SSH HOST KEY KEYS-----                                                                                                                                                                                                                                                                                                                                                                                                          
ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBMHh+hGY/gB19cRuab9T9OAfDGvQS0YMtI9S1yh79WVnPugmOlDPx7oD0RinCBYmNOwKhgSqm27fNTR81szQHzU= root@ip-10-110-197-191                                                                                                                                                                                                                                                    
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIP3iqPWDfS4inLfwNwBX/RfRw1mGCCvbTxyxz0MZ6dRh root@ip-10-110-197-191                                                                                                                                                                                                                                                                                                                                    
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQCyurVyeLhHjeH5TE8RocHyBiXqVGjSPJOJRWag+vlv9lK5LjAeRVmD9pSP5r8yCEuEovDXeEsZkC7PQzEo6DOHNywymybBIDShQ7W/yNKI9L/wdtlG+4y/m/Av/CZdBxS4SyOL/ivTqOGMJjXDj2Q8k3fC1yUmfCo92h/0qxb21zh46uFKHDyAOqHnB7xz2s4mh5ntVxWcGCev5l3R7rJ8x2QCvKGCoNCnBlXvMZQCSUrUUYfkXWqpMBhAyPBL8amJTpi30S+AFt1iU1ujV/6Peqy3qXkQTfOkkbLcBL8KKNk1EmEIFPCKNBU8k76/xCXrND8Fs0pR6rQ3tXBMnFaM+8UPaqYM/Hh7Z3Wk4XaDtQXHfkjRnMbUz1ogIZx3hzmhNAk
jQZYiU3qzrXvdtRaemElmVw0iLndUZr+xZThJPwxBCcAgTRf4ZsPoV4badUMK4IrGqvpjb9MdF44UzH2t/IuVjRxKmi82NCi7CZiBQLq+scCgsd2menKyVeLEKV8= root@ip-10-110-197-191                                                                                                                                                                                                                                                                                       
-----END SSH HOST KEY KEYS-----                                                                                                                                                                                                                                                                                                                                                                                                            
[   68.577134] cloud-init[8522]: Cloud-init v. 22.2-delphix-2022.07.27.05 running 'modules:final' at Mon, 07 Nov 2022 17:15:04 +0000. Up 68.41 seconds.                                                                                                                                                                                                                                                                                    
[   68.577365] cloud-init[8522]: Cloud-init v. 22.2-delphix-2022.07.27.05 finished at Mon, 07 Nov 2022 17:15:04 +0000. Datasource DataSourceEc2Local.  Up 68.57 seconds                                                                                                                                                                                                                                                                    
[  OK  ] Finished Execute cloud user/final scripts.                                                                                                                                                                                                                                                                                                                                                                                        
[  OK  ] Reached target Cloud-init target.                                                                                                                                                                                                                                                                                                                                                                                                 
[  OK  ] Started Delphix osadmin service.                                                                                                                                                                                                                                                                                                                                                                                                  
[  OK  ] Started Delphix Perfomance Statistics.                                                                                                                                                                                                                                                                                                                                                                                            
                                                                                                          
  Delphix Version 7.0.0.0                                      www.delphix.com
┌──────────────────────────────────────────────────────────────────────────┐  
  │ Keyboard layout (press 9 to change): us                                  │                            
  │ Press any other key for console access.                                  │
  │    ┌────────────────────────────────────────────────────────────────┐    │
  │    │                                                                │    │
  │    │    STATUS                     SERVICE                          │    │                            
  │    │    ─────────────────────────────────────────────               │    │                            
  │    │    active     Delphix "rpool" Upgrade Service                  │    │
  │    │    active     Delphix Appliance Platform Service               │    │
  │    │    active     Delphix osadmin service                          │    │
  │    │    active     Delphix Perfomance Statistics                    │    │                            
  │    │    activating Delphix default PostgreSQL database server       │    │                                                                                                                                                                                                                                                                                                                                                             
  │    │    active     Delphix Telegraf Metric Collection Agent         │    │
  │    │    active     Delphix Nginx reverse proxy                      │    │           
  │    │                                                                │    │                            
  │    │    [  OK  ] Started Delphix default PostgreSQL database server.│    │                                                                                                                                       
[  OK  ] Stopped Wait for Network to be Configured.─────────────────────┘    │                                                                                                                                       
         Stopping Wait for Network to be Configured...                       │                                                                                                                                       
         Stopping Network Service...                                         │                                                                                                                                       
[  OK  ] Stopped Network Service.────────────────────────────────────────────┘                            
         Starting Network Service...                                                         
[  OK  ] Started Network Service.                                             
[  OK  ] Started ntp-systemd-netif.service.                                                                                                                                                                          
         Starting Wait for Network to be Configured...                                                    
[  OK  ] Finished Wait for Network to be Configured.                                                                                                                                                                                                                                                                                                                                                                                       
[  OK  ] Started ntp-systemd-netif.service.                                              
         Starting Hostname Service...                                         
[  OK  ] Started Hostname Service.  
...
  Delphix Version 7.0.0.0                                      www.delphix.com                                                                                                                                                                                                                                                                                                                                                             
┌──────────────────────────────────────────────────────────────────────────┐                                                                                                                                                                                                                                                                                                                                                               
  │ Keyboard layout (press 9 to change): us                                  │                                                                                                                                                                                                                                                                                                                                                             
  │ Press any other key for console access.                                  │                                                                                                                                                                                                                                                                                                                                                             
  │    ┌────────────────────────────────────────────────────────────────┐    │                                                                                                                                                                                                                                                                                                                                                             
  │    │                                                                │    │                                                                                                                                                                                                                                                                                                                                                             
  │    │    STATUS                     SERVICE                          │    │                                                                                                                                                                                                                                                                                                                                                             
  │    │    ─────────────────────────────────────────────               │    │                                                                                                                                                                                                                                                                                                                                                             
  │    │    active     Delphix "rpool" Upgrade Service                  │    │                                                                                                                                                                                                                                                                                                                                                             
  │    │    active     Delphix Appliance Platform Service               │    │                                                                                                                                                                                                                                                                                                                                                             
  │    │    active     Delphix osadmin service                          │    │                            
  │    │    active     Delphix Perfomance Statistics                    │    │
  │    │    active     Delphix default PostgreSQL database server       │    │
  │    │    active     Delphix management service                       │    │                            
  │    │    active     Delphix Telegraf Metric Collection Agent         │    │
  │    │    active     Delphix Nginx reverse proxy                      │    │
  │    │                                                                │    │                            
  │    └────────────────────────────────────────────────────────────────┘    │                            
  │  Host: https://ip-10-110-197-191                                         │                            
  │  IPs: 10.110.197.191                                                     │                            
  └──────────────────────────────────────────────────────────────────────────┘   ```

prakashsurya commented:
- `performance_dose_zfs_local` is [here](http://selfservice.jenkins.delphix.com/job/blackbox-self-service/56169/consoleFull)
